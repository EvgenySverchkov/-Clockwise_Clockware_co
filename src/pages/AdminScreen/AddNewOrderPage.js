import React, { useEffect, useContext } from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import { townsInit } from "../../store/townsManagement/actions";
import { addSuitableMasters } from "../../store/masterManagement/actions";
import {
  addCurrentOrderToState,
  changeAdminOrderFormIsLoad,
} from "../../store/ordersManagement/actions";

import Context from "../../ContextComponent";

import { SERVERDOMAIN } from "../../services/serverUrls";

import OrderForm from "../../forms/OrderForm";

function AddNewOrderPage({ history }) {
  const context = useContext(Context);
  const state = useSelector((state) => {
    return {
      currentOrder: state.ordersReducer.currentOrder,
      townsArr: state.townReduser.towns,
      orderFormIsLoad: state.ordersReducer.adminOrderFormIsLoad,
    };
  });
  const dispatch = useDispatch();

  useEffect(function () {
    getTownsFromServerToState();
  }, []);
  function changeHandler(e) {
    let idx = e.target.name;
    dispatch(
      addCurrentOrderToState({
        ...state.currentOrder,
        [idx]: e.target.value,
      })
    );
  }
  function getTownsFromServerToState() {
    const headers = {
      Authorization: "Bearer " + sessionStorage.getItem("token") || "",
      "Content-Type": "application/json",
    };
    fetch(`${SERVERDOMAIN}/towns`, { headers })
      .then((json) => json.json())
      .then((data) => dispatch(townsInit(data)));
  }
  function submitHandler(e) {
    e.preventDefault();
    let trgElem = e.target;
    if (
      !trgElem.town.value ||
      !trgElem.size.value ||
      !trgElem.name.value ||
      !trgElem.email.value ||
      !trgElem.time.value ||
      !trgElem.date.value
    ) {
      context.openErrorWindowWithMsg("Please, fill all fields!!!");
      return false;
    }
    if (trgElem.name.value.length <= 3) {
      context.openWarningTooltip("Name field must be at least 3 characters!", trgElem.name.id);
      return false;
    }
    if (trgElem.name.value.match(/\d/) || !trgElem.name.value.match(/\b\w{3,20}\b/)) {
      context.openWarningTooltip("String name should:\n1. Not contain numbers\n2. Not be shorter than 3 characters\n3. Not longer than 20 characters\n4. Do not contain Cyrillic characters!", trgElem.name.id);
      return false;
    }

    if (!trgElem.email.value.match(/^\w+@[a-zA-Z_0-9]+?\.[a-zA-Z]{2,}$/)) {
      context.openWarningTooltip("Invalid email format!", trgElem.email.id);
      return false;
    }

    if (trgElem.time.value >= "18:00" || trgElem.time.value < "09:00") {
      context.openWarningTooltip("Time should not be more than 18:00 and less than 09:00", trgElem.time.id);
      return false;
    }

    if (!isClientDateLargeThenCurrDate(trgElem.date.value)) {
      context.openWarningTooltip("Date must not be less than or equal to the current date", trgElem.date.id);
      return false;
    }

    let endOrderTime;
    let clientTimeHour = +state.currentOrder.time.match(/\d\d/);
    let clientTimeMin = state.currentOrder.time.match(/:\d\d$/);
    switch (state.currentOrder.size) {
      case "small":
        endOrderTime = clientTimeHour + 1 + clientTimeMin;
        break;
      case "middle":
        endOrderTime = clientTimeHour + 2 + clientTimeMin;
        break;
      case "large":
        endOrderTime = clientTimeHour + 3 + clientTimeMin;
        break;
      default:
        endOrderTime = 0;
    }
    dispatch(changeAdminOrderFormIsLoad(true));
    getFreeMastersByClientTownFromServer(
      SERVERDOMAIN,
      trgElem.town.value,
      state.currentOrder.time,
      endOrderTime,
      state.currentOrder.date
    ).then((data) => {
      dispatch(changeAdminOrderFormIsLoad(false));
      if (data.success) {
        dispatch(addSuitableMasters(data.payload));
        history.push("/admin/freeMasters");
      } else {
        dispatch(addSuitableMasters([]));
        context.openErrorWindowWithMsg(data.msg)
      }
    });
  }
  function getFreeMastersByClientTownFromServer(
    url,
    clientTown,
    clientTimeStart,
    clientTimeEnd,
    clientDate
  ) {
    let obj = {
      town: clientTown,
      timeStart: clientTimeStart,
      timeEnd: clientTimeEnd,
      date: clientDate,
    };
    return fetch(`${url}/masters`, {
      method: "POST",
      headers: {
        Authorization: sessionStorage.getItem("token")
          ? "Bearer " + sessionStorage.getItem("token")
          : "",
        "Content-Type": "application/json;charset=utf-8",
        include: "free",
      },
      body: JSON.stringify(obj),
    }).then((json) => json.json());
  }
  function isClientDateLargeThenCurrDate(clientDate) {
    const clientDt = new Date(clientDate);
    const currDate = new Date();

    if (currDate.getTime() > clientDt.getTime()) {
      return false;
    } else {
      return true;
    }
  }
  return (
    <OrderForm
      currentOrder={state.currentOrder || []}
      changeHandler={changeHandler}
      townsArr={state.townsArr}
      submitHandler={submitHandler}
      isLoadOrderForm={state.orderFormIsLoad}
    />
  );
}

AddNewOrderPage.propTypes = {
  submitHandler: PropTypes.func,
};

export default AddNewOrderPage;
