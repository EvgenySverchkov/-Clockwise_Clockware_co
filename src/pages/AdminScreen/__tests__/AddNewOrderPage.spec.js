import React from "react";
import {create, act} from "react-test-renderer";
import {Provider} from "react-redux";
import configMockStore from "redux-mock-store";

import Context from "../../../ContextComponent";
import AddNewOrderPage from "../AddNewOrderPage";

const mockStore = configMockStore();

const constantDate = new Date('2020-06-13')

Date = class extends Date {
  constructor() {
    return constantDate
  }
}

describe("Test of <AddNewOrderPage />", ()=>{
    global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve())}));
    const initialStore = {
        ordersReducer: {
            currentOrder: {},
            adminOrderFormIsLoad: false,
        },
        townReduser: {
            towns: [{id: 1, name: "Name"}]
        }
    };
    
    const mockContextValue = {
        openErrorWindowWithMsg: jest.fn(),
        openWarningTooltip: jest.fn(),
        warningTooltipMsg: "Message"
    };
    const mockProps = {
        history: {
            push: jest.fn()
        }
    };
    it("Default view", async()=>{
        const store = mockStore(initialStore);
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddNewOrderPage {...mockProps}/>
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });
    
    it("The view after click on submit button and waiting response from server", async()=>{
        const store = mockStore({...initialStore, ordersReducer: {adminOrderFormIsLoad: true}});
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddNewOrderPage {...mockProps}/>
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("The view when towns array is empty", async()=>{
        const store = mockStore({...initialStore, townReduser: {towns: []}});
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddNewOrderPage {...mockProps}/>
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });
});