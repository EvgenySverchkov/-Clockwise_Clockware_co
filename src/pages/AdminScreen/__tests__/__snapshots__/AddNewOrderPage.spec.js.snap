// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <AddNewOrderPage /> Default view 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter 
      name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="email"
    >
      Enter 
      email
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose size of clock
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="smallSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="small"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="smallSize"
      >
        Small
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="middleSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="middle"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="middleSize"
      >
        Middle
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="largeSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="large"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="largeSize"
      >
        Large
      </label>
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose town
    </div>
    <div
      className="form-check-inline"
    >
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          id="Name"
          name="town"
          onChange={[Function]}
          type="radio"
          value="Name"
        />
        <label
          className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
          htmlFor="Name"
        >
          Name
        </label>
      </div>
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose date and time
      <br />
      <sub>
        *time from 8 to 18
      </sub>
    </div>
    <input
      aria-describedby={null}
      className="mr-1"
      id="date"
      max="2021-09-30"
      min="2020-09-30"
      name="date"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="date"
      value=""
    />
    <input
      aria-describedby={null}
      className=""
      id="time"
      max={null}
      min={null}
      name="time"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="time"
      value=""
    />
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Next step"
    />
  </div>
</form>
`;

exports[`Test of <AddNewOrderPage /> The view after click on submit button and waiting response from server 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter 
      name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="email"
    >
      Enter 
      email
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose size of clock
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="smallSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="small"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="smallSize"
      >
        Small
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="middleSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="middle"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="middleSize"
      >
        Middle
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="largeSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="large"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="largeSize"
      >
        Large
      </label>
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose town
    </div>
    <div
      className="form-check-inline"
    >
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          id="Name"
          name="town"
          onChange={[Function]}
          type="radio"
          value="Name"
        />
        <label
          className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
          htmlFor="Name"
        >
          Name
        </label>
      </div>
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose date and time
      <br />
      <sub>
        *time from 8 to 18
      </sub>
    </div>
    <input
      aria-describedby={null}
      className="mr-1"
      id="date"
      max="2021-09-30"
      min="2020-09-30"
      name="date"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="date"
      value=""
    />
    <input
      aria-describedby={null}
      className=""
      id="time"
      max={null}
      min={null}
      name="time"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="time"
      value=""
    />
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Loading..."
    />
  </div>
</form>
`;

exports[`Test of <AddNewOrderPage /> The view when towns array is empty 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter 
      name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="email"
    >
      Enter 
      email
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose size of clock
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="smallSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="small"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="smallSize"
      >
        Small
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="middleSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="middle"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="middleSize"
      >
        Middle
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        className="form-check-input"
        id="largeSize"
        name="size"
        onChange={[Function]}
        type="radio"
        value="large"
      />
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
        htmlFor="largeSize"
      >
        Large
      </label>
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose town
    </div>
    -- List is empty --
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose date and time
      <br />
      <sub>
        *time from 8 to 18
      </sub>
    </div>
    <input
      aria-describedby={null}
      className="mr-1"
      id="date"
      max="2021-09-30"
      min="2020-09-30"
      name="date"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="date"
      value=""
    />
    <input
      aria-describedby={null}
      className=""
      id="time"
      max={null}
      min={null}
      name="time"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      required={true}
      title="Message"
      type="time"
      value=""
    />
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Next step"
    />
  </div>
</form>
`;
